src/v/storage/segment_reader.cc

```diff
@@ -56,7 +56,8 @@ segment_reader::segment_reader(segment_reader&& rhs) noexcept
   , _file_size(rhs._file_size)
   , _buffer_size(rhs._buffer_size)
   , _read_ahead(rhs._read_ahead)
-  , _sanitize(rhs._sanitize) {
+  , _sanitize(rhs._sanitize)
+  , _gate(std::move(rhs._gate)) {
     for (auto& i : _streams) {
```

Comment:
@VladLazar just skimming through some old PR github notifications.

This `gate` move looks scary. IIRC gate guards hold a raw pointer to underlying gate. A move would mean that any active gates would have dangling pointers.

Any concern here? How do we know we are safe?

Reply:
Very good point. I think you're right and this could be problematic. I guess we just don't move `segment_reader` instances while there outstanding ops on them, but there's nothing enforcing this as far as I can tell. I think this was problematic even before the gate was added.

We could wrap the gate in a `unique_ptr` to ensure it outlives the holders, but it still means futures may operate on a moved from `segment_reader` which I think could still cause problems. Maybe we could wrap the whole `segment_reader` into a `unique_ptr` instead. What do you think?

Reply:
>whole segment_reader into a unique_ptr instead. What do you think?

yes, this is the first approach I'd look at.